buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionDownloadPlugin.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle.internal;
+
+import org.elasticsearch.gradle.Architecture;
+import org.elasticsearch.gradle.BwcVersions;
+import org.elasticsearch.gradle.DistributionDownloadPlugin;
+import org.elasticsearch.gradle.DistributionResolution;
+import org.elasticsearch.gradle.ElasticsearchDistribution;
+import org.elasticsearch.gradle.Version;
+import org.elasticsearch.gradle.VersionProperties;
+import org.elasticsearch.gradle.info.BuildParams;
+import org.elasticsearch.gradle.info.GlobalBuildInfoPlugin;
+import org.gradle.api.GradleException;
+import org.gradle.api.NamedDomainObjectContainer;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+
+import static org.elasticsearch.gradle.util.GradleUtils.projectDependency;
+
+/**
+ * An internal elasticsearch build plugin that registers additional
+ * distribution resolution strategies to the 'elasticsearch.download-distribution' plugin
+ * to resolve distributions from a local snapshot or a locally built bwc snapshot.
+ */
+public class InternalDistributionDownloadPlugin implements Plugin<Project> {
+
+    private BwcVersions bwcVersions = null;
+
+    @Override
+    public void apply(Project project) {
+        // this is needed for isInternal
+        project.getRootProject().getPluginManager().apply(GlobalBuildInfoPlugin.class);
+        if (!BuildParams.isInternal()) {
+            throw new GradleException(
```

Comment:
Nice. At least for now until we physically separate this code such that we don't publish internal stuff this seems like a good pattern to ensure that folks don't "accidentally" leverage internal plugins.